ldap/servers/slapd/back-ldbm/db-mdb/mdb_import_threads.c

```diff
@@ -900,24 +910,46 @@ dbmdb_import_entry_info_by_param(EntryInfoParam_t *param, WorkerQueueData_t *wqe
         }
     }
     if (DNRC_IS_ENTRY(dnrc)) {
-        /* Note: If someday we needs to rebuild the entry dn, we could store it in the entry info
-         *  and build it from parent entry info + entry rdn. But as of today we do not need it.
-         */
-        len = strlen(nrdn) + strlen(rdn) + 2 + (5 + pinfo[1]) * sizeof(ID);
+        size_t rdnlen = strlen(rdn);
+        size_t dnlen = 0;
+        if (param->flags & EIP_RDN) {
+            /* In reindex case, dn must be rebuilt to be able to scope properly the vlv index */
+            dnlen = rdnlen + 1 + pinfo[4];  /* dn len (including final \0) */
+        }
+
+        len = rdnlen + strlen(nrdn) + 2 + dnlen + (INFO_IDX_ANCESTORS + 1 + pinfo[INFO_IDX_NB_ANCESTORS]) * sizeof(ID);
         data.mv_data = wqelmt->entry_info = (ID*)slapi_ch_calloc(ALIGN_TO_ID_SLOT(len), sizeof(ID));
         data.mv_size = len;
-        wqelmt->entry_info[0] = param->eid;
-        wqelmt->entry_info[1] = pinfo[1] + 1;
-        wqelmt->entry_info[2] = strlen(nrdn)+1;
-        wqelmt->entry_info[3] = strlen(rdn)+1;
-        if (pinfo[1]) {
+        wqelmt->entry_info[INFO_IDX_ENTRY_ID] = param->eid;
+        if (pinfo[INFO_IDX_ENTRY_ID] == 0) {
+            wqelmt->entry_info[INFO_IDX_NB_ANCESTORS] = 0;
+        } else {
+            wqelmt->entry_info[INFO_IDX_NB_ANCESTORS] = pinfo[INFO_IDX_NB_ANCESTORS] + 1; /* parent ancestors + parent id */
+        }
+        wqelmt->entry_info[INFO_IDX_NRDN_LEN] = strlen(nrdn)+1;
+        wqelmt->entry_info[INFO_IDX_RDN_LEN] = rdnlen+1;
+        wqelmt->entry_info[INFO_IDX_DN_LEN] = dnlen;
+        if (pinfo[INFO_IDX_NB_ANCESTORS]) {
             /* Copy parent ancestors */
-            memcpy(&wqelmt->entry_info[4], &pinfo[4], pinfo[1]*sizeof(ID));
+            memcpy(&wqelmt->entry_info[INFO_IDX_ANCESTORS], &pinfo[INFO_IDX_ANCESTORS], pinfo[INFO_IDX_NB_ANCESTORS]*sizeof(ID));
+        }
+        if (pinfo[INFO_IDX_ENTRY_ID] != 0) {
             /* Then add parent id to ancestors */
-            wqelmt->entry_info[4+pinfo[1]] = pinfo[0];
+            wqelmt->entry_info[INFO_IDX_ANCESTORS+pinfo[INFO_IDX_NB_ANCESTORS]] = pinfo[INFO_IDX_ENTRY_ID];
+        }
+        memcpy(INFO_NRDN(wqelmt->entry_info), nrdn, wqelmt->entry_info[INFO_IDX_NRDN_LEN]);
+        memcpy(INFO_RDN(wqelmt->entry_info), rdn, wqelmt->entry_info[INFO_IDX_RDN_LEN]);
+        if (dnlen>0) {
+            char *dn = INFO_DN(wqelmt->entry_info);
+            memcpy(dn, rdn, rdnlen);
+            if (pinfo[INFO_IDX_DN_LEN]) {
+                dn[rdnlen] = ',';
+                memcpy(dn+rdnlen+1, INFO_DN(pinfo), pinfo[INFO_IDX_DN_LEN]);
```

Comment:
Code quite difficult to understand, sorry in advance if my question is confusing.
Few lines above you computed 'dnlen' with +1 and I understood it was for '\0'. Here we are adding ',' between 'rdn' and '<rest_of_dn>'. I am concerned if that added  ',' was already account in the dnlen.

Reply:
Both the space for ',' and final \0 are accounted in  dnlen:
            dnlen = rdnlen + 1 + pinfo[4];  /* dn len (including final \0) */
in other words:   strlen(rdb) + 1 (for the comma) + parent dn len (including the \0)
 
  BTW there is a minor glitch: should better be INFO_IDX_DN_LEN instead of 4:
       dnlen = rdnlen + 1 + pinfo[INFO_IDX_DN_LEN];